### ISSUE TITLE
AWS ElasticForwarder does not push logs when the Lambda closes due to a raised Exception

### ISSUE THREAD

**Issue by DanInacio** [elastic/elastic-serverless-forwarder#151]
Hi

We have an application hosted in AWS that receives JSON Requests and sends us a JSON Response with content we want to see in ElasticSearch. We started using AWS ElasticForwarder and came across a peculiar situation:

- When the request has the right format (and a response is sent), the ElasticForwarder is called and logs from CloudWatch are seen in ElasticSearch
BUT
- When the request is wrong (differently-named key, for example), raising an Exception, logs from CloudWatch show that an exception is raised and ElasticForwarder also throws an exception.

This results in having ElasticSearch only logging the requests that were correct, and all the errors become invisible from Elastic's point of view.
To reproduce, simply raise a KeyError Exception on the application that you want to track logs from.

Below you will find the AWS CloudWatch logs, from both our hosted application and the ElasticForwarder for a good and bad request.

----------------------------------------------------------
**Application Logs (Bad): Raises an Exception**
<img width="1319" alt="Screen Shot 2022-08-11 at 09 56 46" src="https://user-images.githubusercontent.com/16949504/184089059-085eb1d1-3030-4f1c-a205-2b5816b5fc47.png">

----------------------------------------------------------
**ElasticForwarder Logs (Bad): Abruptly stops running**
<img width="1317" alt="Screen Shot 2022-08-11 at 09 51 36" src="https://user-images.githubusercontent.com/16949504/184088816-257600e7-fad0-44a0-bfe9-e4ce15223592.png">

And the exception details:
```
{
    "@timestamp": "2022-08-11T07:37:12.607Z",
    "log.level": "error",
    "message": "exception raised",
    "ecs": {
        "version": "1.6.0"
    },
    "error": {
        "message": "string argument should contain only ASCII characters",
        "stack_trace": "  File \"/var/task/handlers/aws/utils.py\", line 84, in wrapper\n    return func(lambda_event, lambda_context)\n  File \"/var/task/handlers/aws/handler.py\", line 133, in lambda_handler\n    for (\n  File \"/var/task/handlers/aws/cloudwatch_logs_trigger.py\", line 105, in _handle_cloudwatch_logs_event\n    for log_event, json_object, ending_offset, starting_offset, newline_length in events:\n  File \"/var/task/storage/payload.py\", line 60, in get_by_lines\n    base64_decoded = base64.b64decode(self._payload, validate=True)\n  File \"/var/lang/lib/python3.9/base64.py\", line 80, in b64decode\n    s = _bytes_from_decode_data(s)\n  File \"/var/lang/lib/python3.9/base64.py\", line 39, in _bytes_from_decode_data\n    raise ValueError('string argument should contain only ASCII characters')\n",
        "type": "ValueError"
    },
    "log": {
        "logger": "root",
        "origin": {
            "file": {
                "line": 109,
                "name": "utils.py"
            },
            "function": "wrapper"
        },
        "original": "exception raised"
    },
    "process": {
        "name": "MainProcess",
        "pid": 9,
        "thread": {
            "id": 140534405121856,
            "name": "MainThread"
        }
    }
}
```
----------------------------------------------------------
**Application Logs (Good): Normal Execution**
<img width="1277" alt="Screen Shot 2022-08-11 at 09 57 47" src="https://user-images.githubusercontent.com/16949504/184095113-a6f5fa96-5634-48ed-a387-ca65e62c86ee.png">

----------------------------------------------------------
**ElasticForwarder Logs (Good): Normal Execution**
<img width="1236" alt="Screen Shot 2022-08-11 at 10 36 07" src="https://user-images.githubusercontent.com/16949504/184095480-7f0d2a23-137b-482d-8f79-14883533b1b2.png">

----------------------------------------------------------
**YAML Serverless App Repo Config File (removed data on the region, account, group naming and secrets...)**
```
inputs:
  - type: "cloudwatch-logs"
    id: "arn:aws:logs:[REGION]:[AWS_ACC_ID]:log-group:/aws/lambda/[LOG_GROUP_NAME]:*"
    outputs:
      - type: "elasticsearch"
        args:
          cloud_id: "arn:aws:secretsmanager:[REGION]:[AWS_ACC_ID]:secret:[SECRET_NAME]:[SECRET_KEY]"
          username: "arn:aws:secretsmanager:[REGION]:[AWS_ACC_ID]:secret:[SECRET_NAME]:[SECRET_KEY]"
          password: "arn:aws:secretsmanager:[REGION]:[AWS_ACC_ID]:secret:[SECRET_NAME]:[SECRET_KEY]"
          es_datastream_name: "logs-generic-default"
          batch_max_actions: 500
          batch_max_bytes: 10485760
```

**Comment by aspacca** [comment]
hello @DanInacio 

what version of the forwarder do you use?

> * When the request is wrong (differently-named key, for example), raising an Exception, logs from CloudWatch show that an exception is raised and ElasticForwarder also throws an exception.

can you explain exactly what's this scenario?

the request you are talking about is the request to your application, and the log entry reports this. is this correct?

this is the stack trace of the forwarder:
```
  File "/var/task/handlers/aws/utils.py", line 84, in wrapper
    return func(lambda_event, lambda_context)
  File "/var/task/handlers/aws/handler.py", line 133, in lambda_handler
    for (
  File "/var/task/handlers/aws/cloudwatch_logs_trigger.py", line 105, in _handle_cloudwatch_logs_event
    for log_event, json_object, ending_offset, starting_offset, newline_length in events:
  File "/var/task/storage/payload.py", line 60, in get_by_lines
    base64_decoded = base64.b64decode(self._payload, validate=True)
  File "/var/lang/lib/python3.9/base64.py", line 80, in b64decode
    s = _bytes_from_decode_data(s)
  File "/var/lang/lib/python3.9/base64.py", line 39, in _bytes_from_decode_data
    raise ValueError('string argument should contain only ASCII characters')
```

Indeed `ValueError` is not expected: https://github.com/elastic/elastic-serverless-forwarder/blob/lambda-v1.2.1/storage/payload.py#L61

This is the case when to `base64.b64decode` are passed non-ASCII characters, that should not be the case since the input is coming from cloudwatch logs and should be valid bas64 encoded error

could you share the content of a cloudwatch logs event payload (the one that a lambda receives: `{"awslogs": {"data": etc etc}}`?


Even if the we catch the exception, it won't solve the issue, having the cloudwatch logs event payload not in the right format

**Comment by aspacca** [comment]
do you know if the error arises when the log event is processed from the original cloudwatch logs input or from the continuing queue?

**Comment by DanInacio** [comment]
hello @aspacca 

ElasticForwarder version is 1.2.1

The scenario you asked about is exactly as you thought: the request is the one made to the application. In both scenarios, good and bad, the logs I showed are the application logs and the ElasticForwarder logs

As for the CloudWatch event payload, I am not sure I understand which log that is. Do you want to see more details on the logs from the lambda function whose logs we want to catch? Or more details on the Serverless Application Repository's Cloudwatch logs?

Regardless, below I have sent more detailed logs on what happens during a good request in the ElasticForwarder Application:

----------------------------------------------------------------------
**1. Trigger**
```
{
    "@timestamp": "2022-08-11T07:37:21.255Z",
    "log.level": "info",
    "message": "trigger",
    "ecs": {
        "version": "1.6.0"
    },
    "log": {
        "logger": "root",
        "origin": {
            "file": {
                "line": 60,
                "name": "handler.py"
            },
            "function": "lambda_handler"
        },
        "original": "trigger"
    },
    "process": {
        "name": "MainProcess",
        "pid": 9,
        "thread": {
            "id": 140534405121856,
            "name": "MainThread"
        }
    },
    "type": "cloudwatch-logs"
}
```
----------------------------------------------------------------------
**2. Read Config File from S3**
```
{
    "@timestamp": "2022-08-11T07:37:21.256Z",
    "log.level": "info",
    "message": "config file",
    "bucket_name": "[NAME_HERE]",
    "ecs": {
        "version": "1.6.0"
    },
    "log": {
        "logger": "root",
        "origin": {
            "file": {
                "line": 213,
                "name": "utils.py"
            },
            "function": "config_yaml_from_s3"
        },
        "original": "config file"
    },
    "object_key": "elastic_forwarder/config.yaml",
    "process": {
        "name": "MainProcess",
        "pid": 9,
        "thread": {
            "id": 140534405121856,
            "name": "MainThread"
        }
    }
}
```
----------------------------------------------------------------------
**3. Set ElasticShipper**
```
{
    "@timestamp": "2022-08-11T07:37:21.703Z",
    "log.level": "info",
    "message": "setting ElasticSearch shipper",
    "ecs": {
        "version": "1.6.0"
    },
    "log": {
        "logger": "root",
        "origin": {
            "file": {
                "line": 162,
                "name": "utils.py"
            },
            "function": "get_shipper_from_input"
        },
        "original": "setting ElasticSearch shipper"
    },
    "process": {
        "name": "MainProcess",
        "pid": 9,
        "thread": {
            "id": 140534405121856,
            "name": "MainThread"
        }
    }
}
```
----------------------------------------------------------------------
**4. GET**
```
{
    "@timestamp": "2022-08-11T07:37:21.743Z",
    "log.level": "warning",
    "message": "GET https://[UNIQUE_ID].[REGION].aws.found.io/ [status:403 request:0.039s]",
    "ecs": {
        "version": "1.6.0"
    },
    "log": {
        "logger": "elasticsearch",
        "origin": {
            "file": {
                "line": 288,
                "name": "base.py"
            },
            "function": "log_request_fail"
        },
        "original": "GET https://[UNIQUE_ID].[REGION].aws.found.io/ [status:403 request:0.039s]"
    },
    "process": {
        "name": "MainProcess",
        "pid": 9,
        "thread": {
            "id": 140534405121856,
            "name": "MainThread"
        }
    }
}
```
----------------------------------------------------------------------
**5. POST**
```
{
    "@timestamp": "2022-08-11T07:37:21.751Z",
    "log.level": "info",
    "message": "POST https://[UNIQUE_ID].[REGION].aws.found.io/_bulk [status:200 request:0.007s]",
    "ecs": {
        "version": "1.6.0"
    },
    "log": {
        "logger": "elasticsearch",
        "origin": {
            "file": {
                "line": 265,
                "name": "base.py"
            },
            "function": "log_request_success"
        },
        "original": "POST https://[UNIQUE_ID].[REGION].aws.found.io/_bulk [status:200 request:0.007s]"
    },
    "process": {
        "name": "MainProcess",
        "pid": 9,
        "thread": {
            "id": 140534405121856,
            "name": "MainThread"
        }
    }
}
```
----------------------------------------------------------------------
**6. Handle ES Outcome**
```
{
    "@timestamp": "2022-08-11T07:37:21.751Z",
    "log.level": "info",
    "message": "elasticsearch shipper",
    "ecs": {
        "version": "1.6.0"
    },
    "failed": 0,
    "log": {
        "logger": "root",
        "origin": {
            "file": {
                "line": 152,
                "name": "es.py"
            },
            "function": "_handle_outcome"
        },
        "original": "elasticsearch shipper"
    },
    "process": {
        "name": "MainProcess",
        "pid": 9,
        "thread": {
            "id": 140534405121856,
            "name": "MainThread"
        }
    },
    "success": 6
}
```
----------------------------------------------------------------------
**7. Output Message**
```
{
    "@timestamp": "2022-08-11T07:37:21.752Z",
    "log.level": "info",
    "message": "lambda processed all the events",
    "ecs": {
        "version": "1.6.0"
    },
    "empty_events": 0,
    "log": {
        "logger": "root",
        "origin": {
            "file": {
                "line": 187,
                "name": "handler.py"
            },
            "function": "lambda_handler"
        },
        "original": "lambda processed all the events"
    },
    "process": {
        "name": "MainProcess",
        "pid": 9,
        "thread": {
            "id": 140534405121856,
            "name": "MainThread"
        }
    },
    "sent_event": 6,
    "skipped_events": 0
}
```

**Comment by aspacca** [comment]
> 4. GET

this is strange, you are getting a 403 forbidden response to a GET request to elaticsearch, it should be logged as `info` with a `status:200`. the POST request is successful indeed

**Comment by aspacca** [comment]
> As for the CloudWatch event payload, I am not sure I understand which log that is. Do you want to see more details on the logs from the lambda function whose logs we want to catch? Or more details on the Serverless Application Repository's Cloudwatch logs?

"CloudWatch Logs invokes your function asynchronously with an event that contains log data. The value of the data field is a Base64-encoded .gzip file archive."

https://docs.aws.amazon.com/lambda/latest/dg/services-cloudwatchlogs.html

```
{
  "awslogs": {
    "data": "..."
  }
}
```

it what's inside `data` that seems to contain non-ASCII characters: that's what triggering the exception in the forwarder, and it should not be the case, since it won't be valida base64 encoded data


you can add the environment variable `LOG_LEVEL=DEBUG` (beware, this make the forwarder very verbose)

if you look for this log entry you will find the payload I was talking about:
https://github.com/elastic/elastic-serverless-forwarder/blob/main/handlers/aws/handler.py#L55

beware that it might contains sensitive data, and you don't want to share it without checking before
on a mac you can decode with the following
```
$ echo "%WHAT'S IN DATA%"|base64 -D|zcat
```

you can also check just if the payload contains non-ASCII chars with the following:
```
$ echo "%WHAT'S IN DATA%"|iconv -f ascii
```

if it returns an error then the payload contains non-ASCII chars

**Comment by DanInacio** [comment]
> > 4. GET
> 
> this is strange, you are getting a 403 forbidden response to a GET request to elasticsearch, it should be logged as `info` with a `status:200`. the POST request is successful indeed
> 
> it's not related to the exception anyway

Right, I forgot to paste the 2 logs that show up between GET and POST:
```
> /var/task/elasticsearch/helpers/actions.py:240: ElasticsearchWarning: The client is unable to verify that the server is Elasticsearch due security privileges on the server side
> resp = client.bulk(*args, body="\n".join(bulk_actions) + "\n", **kwargs)
```

Since it still POSTs anyway, I agree, it should not be related to the Exception.

**Comment by DanInacio** [comment]
As for the encoded .gzip file archives, I will take a look at them and let you know more tomorrow.

**Comment by DanInacio** [comment]
Hi Andrea

To keep you updated, we have yet to find these logs containing
```
{
  "awslogs": {
    "data": "..."
  }
}
```
Retrieving logs with the command _get-log-events_ only returns the message, timestamp and ingestion time. Do you know of an alternative way to obtain these encoded logs?

**Comment by aspacca** [comment]
hi @DanInacio 

> Retrieving logs with the command _get-log-events_ only returns the message, timestamp and ingestion time. Do you know of an alternative way to obtain these encoded logs?

yes, you have to add the environment variable `LOG_LEVEL=DEBUG` to the forwarder and look of the log entry with the following payload in the logs of the forwarder
```
"lambda triggered", extra={"event": lambda_event, "invoked_function_arn": lambda_context.invoked_function_arn}
```

**Comment by DanInacio** [comment]
Hi Andrea

Below follows an example of one of our awslogs' data string. It is important to note that this string's content has been changed to remove any sensitive data. However, I base64.decoded the original awslog locally and it was successfully decoded. Also checked via the cli commands you provided and no non-ASCII characters exist within our logs:

'eyJtZXNzYWdlVHlwZSI6ICJEQVRBX01FU1NBR0UiLCAib3duZXIiOiAiMTExIiwgImxvZ0dyb3VwIjogIi9hd3MvbGFtYmRhL2R1bW15X2dyb3VwIiwgImxvZ1N0cmVhbSI6ICJkdW1teV9zdHJlYW0iLCAic3Vic2NyaXB0aW9uRmlsdGVycyI6IFsic2VydmVybGVzc3JlcG8tZWxhc3RpYy1zZXJ2ZXJsZXNzLWZvcndhcmRlci1FbGFzdGljU2VydmVybGVzc0ZvcndhcmRlckFwcGxpY2F0aW9uLTE3VDY0TkxOSEdHNE4tQXBwbGljYXRpb25FbGFzdGljU2VydmVybGVzc0ZvcndhcmRlckNsb3VkV2F0Y2hMb2dzRXZlbnRubUFuSkZsaEdYLU44VEZVU1Y2OTRMWCJdLCAibG9nRXZlbnRzIjogW3siaWQiOiAiMzcwMzMzOTk4NDI1OTU2ODA5ODY2MjY2OTE5NTI2MzcwMDM0MDcwNTY5NDQwOTE1NTQ2NDM5NjgiLCAidGltZXN0YW1wIjogMTY2MDYzNTA4MjUwMSwgIm1lc3NhZ2UiOiAiU1RBUlQgUmVxdWVzdElkOiBjYTZkNjljYy03NTc3LTQ3YWUtOGMxOC02OWIzYTJkNjkwMTIgVmVyc2lvbjogJExBVEVTVFxuIn0sIHsiaWQiOiAiMzcwMzMzOTk4NDI2NDAyODI0NzcwMjM3NTMxOTg5MjAwNzQ4NDM2MDIyNDA4MTQ1NjY2MDQ4MDEiLCAidGltZXN0YW1wIjogMTY2MDYzNTA4MjUwMywgIm1lc3NhZ2UiOiAie1wiQHRpbWVzdGFtcFwiOlwiMjAyMi0wOC0xNlQwNzozMToyMi41MDNaXCIsXCJsb2cubGV2ZWxcIjpcImVycm9yXCIsXCJtZXNzYWdlXCI6XCJJbnZhbGlkL2luY29tcGxldGUgYXJndW1lbnRzOiB7J2VuZHBvaW50X25hbWUnOiAnZHVtbXlfbmFtZScsICdzdXBwbGllcl9pZCc6ICcxMTExMTEnLCAnc3VwcGxpZXJfbmFtJzogJ2R1bW15IG5hbWUnLCAnZG9jdW1lbnRfbGFuZ3VhZ2UnOiAxLCAnc2RzX2ZpbGVfcGF0aCc6ICdodHRwczovL2xpbmtfaGVyZSd9XCIsXCJlY3NcIjp7XCJ2ZXJzaW9uXCI6XCIxLjYuMFwifSxcImxvZ1wiOntcImxvZ2dlclwiOlwicm9vdFwiLFwib3JpZ2luXCI6e1wiZmlsZVwiOntcImxpbmVcIjo4MyxcIm5hbWVcIjpcImZ1bmN0aW9uLnB5XCJ9LFwiZnVuY3Rpb25cIjpcImhhbmRsZXJcIn0sXCJvcmlnaW5hbFwiOlwiSW52YWxpZC9pbmNvbXBsZXRlIGFyZ3VtZW50czogeydlbmRwb2ludF9uYW1lJzogJ2R1bW15LW5hbWUnLCAnc3VwcGxpZXJfaWQnOiAnMTExMScsICdzdXBwbGllcl9uYW0nOiAnZHVtbXlfc3VwcGxpZXInLCAnZG9jdW1lbnRfbGFuZ3VhZ2UnOiAxLCAnc2RzX2ZpbGVfcGF0aCc6ICdodHRwczovL2xpbmtfaGVyZSd9XCJ9LFwicHJvY2Vzc1wiOntcIm5hbWVcIjpcIk1haW5Qcm9jZXNzXCIsXCJwaWRcIjo4LFwidGhyZWFkXCI6e1wiaWRcIjoxNDA2NTc2MjA5NjkyODAsXCJuYW1lXCI6XCJNYWluVGhyZWFkXCJ9fX1cbiJ9LCB7ImlkIjogIjM3MDMzMzk5ODQyNjQwMjgyNDc3MDIzNzUzMTk4OTIwMDc0ODQzNjAyMjQwODE0NTY2NjA0ODAyIiwgInRpbWVzdGFtcCI6IDE2NjA2MzUwODI1MDMsICJtZXNzYWdlIjogIltFUlJPUl0gS2V5RXJyb3I6ICdpbnZhbGlkX2xhbWJkYV9pbnB1dF9wYXJhbWV0ZXJzJ1xyVHJhY2ViYWNrIChtb3N0IHJlY2VudCBjYWxsIGxhc3QpOlxyXHUwMGEwXHUwMGEwRmlsZSBcIi92YXIvdGFzay9mdW5jdGlvbi5weVwiLCBsaW5lIDg0LCBpbiBoYW5kbGVyXHJcdTAwYTBcdTAwYTBcdTAwYTBcdTAwYTByYWlzZSBLZXlFcnJvcihJTlZBTElEX0FSR1NfRVJST1IpXG4ifSwgeyJpZCI6ICIzNzAzMzM5OTg0MjY4NDg4Mzk2NzQyMDgxNDQ0NTIwMzE0NjI4MDE0NzUzNzUzNzU3ODU2NTYzNSIsICJ0aW1lc3RhbXAiOiAxNjYwNjM1MDgyNTA1LCAibWVzc2FnZSI6ICJFTkQgUmVxdWVzdElkOiBjYTZkNjljYy03NTc3LTQ3YWUtOGMxOC02OWIzYTJkNjkwMTJcbiJ9LCB7ImlkIjogIjM3MDMzMzk5ODQyNjg0ODgzOTY3NDIwODE0NDQ1MjAzMTQ2MjgwMTQ3NTM3NTM3NTc4NTY1NjM2IiwgInRpbWVzdGFtcCI6IDE2NjA2MzUwODI1MDUsICJtZXNzYWdlIjogIlJFUE9SVCBSZXF1ZXN0SWQ6IGNhNmQ2OWNjLTc1NzctNDdhZS04YzE4LTY5YjNhMmQ2OTAxMlx0RHVyYXRpb246IDIuMzcgbXNcdEJpbGxlZCBEdXJhdGlvbjogMyBtc1x0TWVtb3J5IFNpemU6IDE4MDAgTUJcdE1heCBNZW1vcnkgVXNlZDogMjkwIE1CXHRcbiJ9XX0='

**Comment by aspacca** [comment]
> Below follows an example of one of our awslogs' data string. It is important to note that this string's content has been changed and re-encoded to remove any sensitive data.

unluckily the base64 re-encoding of the amended content is not significative to reproduce the bug :)

it was helpful anyway because I was able to see that the the payload contains`\xa0`, that's indeed LATIN-1 and not ASCII

this lead to me to identify the fact that the exception is not raised on decoding `{"awslogs":{"data":""}}` but on decoding `{"logEvents": [{"message":""}]}`

I will add a fix on 1.3.0 release

**Comment by DanInacio** [comment]
Hi Andrea

I can confirm v1.3.0 has fixed this issue and incorrect messages with this encoding are properly forwarded to Elastic :)
